mmselfsup/models/algorithms/interclr_moco.py

```diff
@@ -0,0 +1,426 @@
+        self.neg_sampling = neg_sampling
+        self.semihard_neg_pool_num = semihard_neg_pool_num
+        self.semieasy_neg_pool_num = semieasy_neg_pool_num
+        self.intra_cos = intra_cos_marign_loss
+        self.intra_cos_margin = intra_cos_margin
+        self.intra_arc = intra_arc_marign_loss
+        self.intra_arc_margin = intra_arc_margin
+        self.inter_cos = inter_cos_marign_loss
+        self.inter_cos_margin = inter_cos_margin
+        self.inter_arc = inter_arc_marign_loss
+        self.inter_arc_margin = inter_arc_margin
+        self.intra_loss_weight = intra_loss_weight
+        self.inter_loss_weight = inter_loss_weight
+        self.share_neck = share_neck
+        self.num_classes = num_classes
+
+        # create the queue
+        self.register_buffer('queue', torch.randn(feat_dim, queue_len))
+        self.queue = nn.functional.normalize(self.queue, dim=0)
+        self.register_buffer('queue_ptr', torch.zeros(1, dtype=torch.long))
+
+    def init_weights(self):
+        """Initialize base_encoder with init_cfg defined in backbone."""
+        super(InterCLRMoCo, self).init_weights()
+
+        for param_q, param_k in zip(self.encoder_q.parameters(),
+                                    self.encoder_k.parameters()):
+            param_k.data.copy_(param_q.data)
+            param_k.requires_grad = False
+        if not self.share_neck:
+            for param_q, param_k in zip(self.inter_neck_q.parameters(),
+                                        self.inter_neck_k.parameters()):
+                param_k.data.copy_(param_q.data)
+                param_k.requires_grad = False
+
+    @torch.no_grad()
+    def _momentum_update_key_encoder(self):
+        """Momentum update of the key encoder."""
+        for param_q, param_k in zip(self.encoder_q.parameters(),
+                                    self.encoder_k.parameters()):
+            param_k.data = param_k.data * self.momentum + \
+                           param_q.data * (1. - self.momentum)
+        if not self.share_neck:
+            for param_q, param_k in zip(self.inter_neck_q.parameters(),
+                                        self.inter_neck_k.parameters()):
+                param_k.data = param_k.data * self.momentum + \
+                               param_q.data * (1. - self.momentum)
+
+    @torch.no_grad()
+    def _dequeue_and_enqueue(self, keys):
+        """Update queue."""
+        # normalize
+        keys = nn.functional.normalize(keys, dim=1)
+        # gather keys before updating queue
+        keys = concat_all_gather(keys)
+
+        batch_size = keys.shape[0]
+
+        ptr = int(self.queue_ptr)
+        assert self.queue_len % batch_size == 0  # for simplicity
+
+        # replace the keys at ptr (dequeue and enqueue)
+        self.queue[:, ptr:ptr + batch_size] = keys.transpose(0, 1)
+        ptr = (ptr + batch_size) % self.queue_len  # move pointer
+
+        self.queue_ptr[0] = ptr
+
+    def contrast_intra(self, q, k):
```

Comment:
can this function be reused in InterCLR-NPID or  InterCLR-BYOL?

Reply:
It cannot be reused as different methods have different forms of loss.